monix-execution/jvm/src/main/scala/monix/execution/misc/Local.scala

```diff
@@ -0,0 +1,104 @@
+package monix.execution.misc
+
+import java.util.UUID
+
+object Local {
+
+  type Context = scala.collection.immutable.Map[String, Option[LocalContext]]
+
+  trait LocalContext
+
+  private[this] def context(k: String, v: Option[LocalContext]): Context =
+    scala.collection.immutable.Map(k -> v)
+
+  private[this] val localContext = ThreadLocal[Context]()
+
+  private def register(): String = synchronized {
+    UUID.randomUUID().toString
+  }
+
+  /**
+    * Return the state of the current Local state.
+    */
+  def getContext(): Context =
+    localContext.get
+
+  /**
+    * Restore the Local state to a given Context.
+    */
+  def setContext(ctx: Context): Unit =
+    localContext.set(ctx)
+
+  /**
+    * Clear the Local state.
+    */
+  def clearContext(): Unit =
+    localContext.set(null)
+
+  /**
+    * Execute a block of code using the specified state of Local
+    * Context and restore the current state when complete
+    */
+  def withContext[T](ctx: Context)(f: => T): T = {
```

Comment:
I care about performance a lot and having this be a function call is probably going to mess up the numbers - just a feeling, I haven't tested anything.

One possibility is for us to not work with it in the implementation and just leave it in the API for the users.
Another possibility would be for this to be implemented by a macro, like we have for `Atomic.transform*` functions.

Well, validation that transforming it into a macro improves performance would be nice. As an example, here is:

1. [syncOnContinue macro](https://github.com/monix/monix/blob/master/monix-execution/shared/src/main/scala/monix/execution/Macros.scala#L50)
2. [signature](https://github.com/monix/monix/blob/master/monix-execution/shared/src/main/scala/monix/execution/Ack.scala#L93)

That one is being defined as an extension method, which I think it is cool for macros.

In other words, I am not sure if turning this `withContext` into a macro helps performance, but if it does, then we definitely need to do it.

Reply:
@alexandru I can leave it as it is and do the macro alternative so we can compare in the future.

Reply:
This PR was sent my way 😄. So I'll weigh in on this:

If you covert this method to a macro, the only difference you'd be able to make is that you can guarantee that it's inlined at call site. I doubt that it's going to make a difference as the method is only called a few times internally. Granted they may be on hot paths-- but the benefit of inlining might be very small, and the JVM might actually do the inlining already.

If we have a way to do benchmarking, it would be trivial to inline this by hand in the few spots it's used.

Reply:
@alexandru yeah, I brought a friend that knows more about macros than I do!

Reply:
😃 happy to hear your opinion @andyscott 

Right, so I agree that these are just hunches and before making changes for performance reasons, we should test that it's actually good for performance, because I discovered multiple times that the JVM defies reason 😏